Initial Playbooks

coreyogburn opened the pull request:
This includes the Playbook module for downloading, updating, and referencing playbooks. There's also the Playbook Handler with the first few endpoints for retrieving and converting playbooks. And most notably this contains the playbook sections in Alerts where the questions can be seen.

coreyogburn committed:
Initial Playbook Backend Work

Playbook Module, Model, and Handler.

The module schedules when to update the playbooks and does so similarly to how the AI Summary repo is maintained.

After retrieval, the module will convert all the question's sigma questions into OQL.

Finally the module organizes the modules by detection, category, and engine. For now, a playbook is dedicated to an engine iff it has no detection public Id or category specified. All playbooks are stored in memory

The handler has 2 basic calls: Get Playbook by Playbook Id, and Get Playbook(s) by Detection Id. Get playbook by Id returns specific playbooks. Get playbooks by detection id finds every detection-specific and category-specific playbook and returns them, if there are none then the engine-specific playbook(s) will be returned.

coreyogburn committed:
Conversion Adjustments

When we have the sigma cli convert a question's query, we need it to convert all the variable field names as well. The sigma cli pipeline has been updated to return a JSON object with the converted query and the new field names. Also, if the yaml document separator `---` is used to supply multiple queries to the sigma cli, multiple JSON objects will be returned. The playbook module has been updated to take advantage of these changes.

Improved logging.

coreyogburn committed:
Ungrouped Alerts

Load playbooks recursively from target directory.

Ensure Playbooks are loaded ASAP with firstRun sync.OnceFunc.

On alerts, in the ungrouped table, begin adding the tabs to the expanded row of an alert.

The playbooks are requested when the "Guided Analysis" tab is clicked. A loading circle displays in the tab body while the playbooks are fetched. Questions+Contexts are rendered in an accordion fashion.

coreyogburn committed:
Visual Cleanup and Bug Fixing

Visually cleaned up the progress so far because it'll be used as a template for other places (grouped alerts).

Only provide quick answers to the questions if the licenseState === 'active' using new $root.isPro.

Behave better around refreshing while rows are expanded. Probably still more to do on this front, but better.

coreyogburn committed:
Tightening Bolts

For aggregate alerts, get the latest result and use it to provide an expanded row of details and playbook info.

Ensure a question with a bad query doesn't stop other questions from attempting to be answered.

Sort and limit aggregate events.

Show timestamps next to non-aggregate Instant Insights.

Ensure certain new code/elements are only active when on the Alerts page.

Style changes.

Ensure hunt is unaffected by these changes intended for the alerts page.

coreyogburn committed:
Small Tweaks

If a question's answers come back empty, the table now reports "No Data"

Timestamps in Instant Insights are no longer bold.

applyEventColumnWidth does nothing if it's only required parameter isn't supplied.

coreyogburn committed:
New Variable Substitution Flow

Conversion was moved from SOC startup to upon request by the UI. Because of things like `startswith` in the query modifying variable values to use wildcards, we can't convert the sigma to OQL before variable substitution. So we no longer convert on startup, we deliver playbooks with unconverted sigma queries, the UI fills in the variables, the filled sigma is sent to the API to convert to OQL.

Changed field casing on the models returned by the API. Updated references in the UI.

New default playbook repo branch.

coreyogburn committed:
Multiple Playbooks, Fewer Errors

Requesting a playbook for a detection now calls ExtractDetails with the right engine so the Category will be filled when looking up playbooks. This now results in more than 1 playbook coming back for some detections.

Playbook question iteration now puts every question in the same v-expansion-panels element producing a seamless list of all questions.

Some backend endpoints now return empty arrays instead of null.

If there's no playbooks and thus no questions, the UI won't request the backend to convert them. If the backend receives 0 queries to convert, it won't attempt to convert them and instead return an empty array.

defensivedepth committed:
Initial Sigma backend support  - OQL

coreyogburn committed:
Minor Tweaks

Show more info on screen. Start adding tests.

coreyogburn committed:
Fixes and Some Tests

Fixed the timezone issues in buildQuestionRange.

Removed some of the overreaching parts of getEventField for now. We'll tighten the playbook in places where the fields are wrong.

New way to retrieve a timestamp from an event that should work on any event given to it (an alert event that has original values under `event_data` or a regular event).

Todo: Move new strings to i18n, more tests.

coreyogburn committed:
Preparing for merge

Fix Alert Timestamp in ungrouped events table.

buildDateRange now respects the configured timezone.

No more sticky expanded row content.

Removed applyEventColumnWidth and instead use the value eventColumnWidth directly.

First pass at doc strings in the playbook handler. May need updating as we think through permissions.

Converted strings to i18n entries.

Added tests.

coreyogburn committed:
Test Coverage

Added Go tests to the playbook module.

Refactored a few bits of playbook to use the IOManager rather than direct IO methods.

Renamed pbm to pdm.

jertel commented on html/css/app.css:
```diff
@@ -1007,4 +1007,48 @@ ul ul ul {
\ No newline at end of file
+}
+
+.italic {
```
Why not use the Vuetify bold, italic, whitespace, etc classes?

coreyogburn replied:
italic and bold have been swapped out for font-italic and font-weight-bold but according to Vuetify's docs (https://vuetifyjs.com/en/styles/text-and-typography/) there's no class for setting `white-space: normal`.

jertel commented on html/js/app.js:
```diff
@@ -1524,6 +1524,9 @@ $(document).ready(function () {
           }
         },
+        isPro() {
+          return this.licenseStatus === LICENSE_STATUS_ACTIVE;
```
Need to avoid checking for "pro" and use license features. See `isLicensed(feat)`. We can add a new `IIS`  (Instant InSight) or whatever we need for this.

coreyogburn replied:
Did some learning. Modified it to use isLicensed and removed isPro. Added a new FEAT_IIS to backend and frontend.

jertel commented on html/js/i18n.js:
```diff
@@ -432,6 +435,7 @@ const i18n = {
       eventCaseTitle: 'Event Escalation from SOC',
       eventExpandHelp: 'Show all event fields',
       eventFetchTook: 'The backend data fetch took ',
+      eventHappenedAt: 'Original event happened at',
```
Need better verbiage than "happened". Perhaps `Originally occurred at`, `Original Event Timestamp` or something like that.

coreyogburn replied:
Done.

jertel commented on model/playbook.go:
```diff
@@ -0,0 +1,33 @@
+package model
+
+type Playbook struct {
+	Id                string      `json:"id"`
```
All of these model fields will need a comment line above them, and an `example:'abc'` added to the json: annotation in order for them to render correctly in the API docs.

coreyogburn replied:
Added.

jertel commented on server/modules/playbook/playbook.go:
```diff
@@ -0,0 +1,435 @@
+	}
+
+	logger.WithFields(log.Fields{
+		"duration": time.Since(start).Seconds(),
```
Change log field to `organizePlaybookDuration` to ensure it doesn't overlap with other log lines.

coreyogburn replied:
Too simple a name, I should've seen this. Changed.

jertel commented on server/modules/playbook/playbook.go:
```diff
@@ -0,0 +1,435 @@
+	}
+
+	logger.WithFields(log.Fields{
+		"duration":      time.Since(start).Seconds(),
```
Change log field to organizePlaybookDuration to ensure it doesn't overlap with other log lines.

coreyogburn replied:
To not collide with the previous request, I renamed it to `readPlaybookDuration`.

coreyogburn committed:
Code Review Fixes

Removed `italic` and `bold` classes in favor of Vuetify's `font-italic` and `font-weight-bold` classes respectively.

Removed $root.isPro in favor of using $root.isLicensed with the new feat FEAT_IIS (instant insights).

Improved i18n strings.

Improved model documentation and log fields. Added logging to a few places that were missing it.

coreyogburn committed:
More Code Review Fixes

Add yaml tags to the Auditable struct and use it for Playbooks. Refactor some Playbook fields to better align with existing models. Certain Auditable fields will go unused until we can edit playbooks.

Update LicenseManager tests to acknowledge the new FEAT_IIS.

Add some missing contexts to certain playbookstore calls.

coreyogburn committed:
Extra Tweaks

Add an iis entry to i18n for listing features on the license page.

The dockerfile copies over a whl file, installs it, and leaves it there. Now the dockerfile will remove the file after installation.
